eclipse-theia/theia
# Implement "headless plugins" in a new plugin host outside of frontend connections

#### What it does

Implement a new kind of plugin that I call "headless plugin" that is hosted in a child Node process like the current backend plugins _except_ that this host is not started on a per-frontend-connection basis but exists independently of all such connections.
Thus it is "headless" as it will never know a Theia frontend.

This implements the functionality requested in #12946 and, because it makes no sense not to, #12959.

There are a few points to note about this implementation:

- I am not wedded to the term "headless plugin." I welcome suggestions for a better or more descriptive name, but I would like to avoid the term "backend" because we already have backend plugins that are the backend of a frontend connection, whereas headless plugins have no frontend. And the resulting similarity of names in the code will just be too confusing
- several of the existing interfaces and classes in the `plugin-ext` package are refactored to pull up common functionality to reuse for headless plugins with the aim of promoting maintainability of both plugin stacks over the long term. I have endeavoured to maintain API compatibility for Theia extenders, but please do keep a sharp eye out for compatibility issues
- ~a `@theia/headless-plugin` API declaration is provided that is essentially a greatly stripped-down variant of `@theia/plugin`, providing only capabilities that are available outside of frontend connections, and probably even less than that (see below for some possible follow-up in that area)~
- it is intended by this PR that the support for "headless plugins" be optional. A Theia application could be built without the `headless-plugin` and `headless-plugin-ext` packages and work exactly as it does today. If the Theia team would rather just fold support for headless plugins into the existing `plugin-ext` package (e.g. because that would eliminate the need for some of the refactorings done in this PR, actually simplifying the code and possibly making it more correct) then I would be happy to do that
- the layout of the source tree in the new `headless-plugin-ext` extension tries to match the organization of corresponding code in the `plugin-ext` package, to make it easier to understand the similarities and differences. However, as there is no frontend to headless plugins, of course there are no `browser` directories so some things (such as `XyzMain` API) are organized differently, usually in the `node/` folders
- refactoring in the Terminal Service Ext API merged after this PR was initially implemented required some late, and quick, adaptation that may not be ideal, such as the `MinimalTerminalServiceExt` type used by the `AbstractPluginManagerExt` class to declare only what it actually needs to do its work

#### How to test

This PR includes new examples demonstrating the use case that motivated issue #12946 that can be used for testing. These are:

- `examples/api-provider-sample/`: a new example Theia extension that demonstrates how to provide a custom API object for plugins to use. This is important because the original use case motivating the headless plugins feature is one in which those plugins exist to plug into such application-specific APIs. The API provides a "Greeting of the Day" service with custom functions to call, classes to create, and call-backs for a plugin to register for events
- `sample-plugins/sample-namespace/plugin-gotd`: an example dual headless/backend plugin that uses the greeting-of-the-day API of the other new example ~and also illustrates simple usage of the API provided by Theia, itself, to headless plugins~. It further demonstrates how Theia plugins may provide both backend and headless entry-points to run in both plugin hosts

To test the intended use case of this feature, simply copy the `plugin-gotd-v1.44.0.tgz` tarball from the sample plugin into your `plugins/` folder and start up the Browser Backend. You should see the new headless plugin host start up, load this plugin, and output from it on the console using `[GOTD]` tags to help you locate them in the output. Then hit the frontend in your browser to see that the connection-scoped backend plugin host does its usual thing, and that the GOTD plugin's backend entry-point runs in this context, printing messages tagged with `[GOTD-BE]` to distinguish them from messages printed by the headless plugin.

Then, of course, test the same behaviour in Electron as much as makes sense.

Apart from that, please go as crazy as you like in regression testing the usual plugin capabilities of Theia.

#### Follow-ups

- this PR does not yet include updates to user guides and other public documentation about the Theia plugin architecture. I will be happy to make such updates in this PR once we have agreed (if we do agree) to merge the implementation. Or I will be equally happy to raise a follow-up issue to tackle the docs
- ~some~ no attempt has been made to support ~a few~ any core Theia services in the headless plugin host that should be available generally, outside of frontend connections. ~I don't really know how to test most of that so there may be problems found by early adopters that we need to fix~
- ~also there may be other core Theia services that should be exposed in the new `@theia/headless-plugin` API that currently are not, contribution points and activation events to support in the package manifest that currently are not implemented. I should like to leave it to adopters to identify real-world needs in these areas that we can add later~
- this PR does not provide any default API (analogous to the `theia` or `vscode` API namespace) to plugins but relies on extensions and other plugins to provide APIs, as demonstrated by the new example GotD plugin and the API that it uses provided by the new example extension. It is an area of future consideration whether to provide some default API namespace and what capabilities it should offer, as use cases are discovered

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

----

### Updated 5 Dec 2023

This description is amended to rebrand the new kind of plugin as "headless plugin" (was "base plugin") and to adapt the test instructions for the now dual nature of the example GotD plugin, which previously supported only headless mode but now also backend (of a frontend connection) mode.

### Updated 11 Dec 2023

This description is amended to reflect the withdrawal of the default API namespace that was provided to headless plugins.
